### ISSUE TITLE
gtk reactor not delivering callbacks timely

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7193]
|<img alt="dektwisted's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| dektwisted reported|
|-|-|
|Trac ID|trac#552|
|Type|defect|
|Created|2004-03-16 19:54:19Z|

<details><summary>Searchable metadata</summary>

```
trac-id__552 552
type__defect defect
reporter__dektwisted dektwisted
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1079466859000000 1079466859000000
changetime__1475187433412649 1475187433412649
version__ 
owner__itamarst itamarst
cc__exarkun cc__itamarst cc__warner cc__dektwisted
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="dektwisted's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dektwisted commented|
|-|-|

```
#!html
<pre>
I've posted this code to thelist and several people confirm that the gtk reactor
is not delivering callback events in a timely fashion.  The wx reactor works
normally, as does the plain reactor.  When run the following could should report 
TestThingy messages at approximately 0.1 second intervals, but instead, it
happens much less frequenctly.  Tested with latest release of Twisted.

import time, sys

from twisted.internet import gtk2reactor
gtk2reactor.install()
from twisted.python import threadable
threadable.init()
from twisted.internet import reactor
from twisted.python import log
log.startLogging(sys.stdout)

class TestThingy:
   def __init__(self):
       print "TestThingy created"

   def run(self):
       print "TestThingy run"
       for i in range(10):
           print "TestThingy message"
           time.sleep(0.1)
       print "TestThingy done"
      class ThingyManager:
   def __init__(self):
       pass
         def submitThingy(self):
       log.err("ThingyManager submitting")
       f = TestThingy()
       log.err("ThingyManager created")
       reactor.callInThread(f.run)
       log.err("ThingyManager calledInThread")

def ticker(*args):
   log.err("ticker: %d" % time.time())
   reactor.callLater(1, ticker)

t = ThingyManager()
reactor.callLater(0, ticker)
reactor.callLater(3, t.submitThingy)
reactor.run()
log.err("Reactor done")
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="dektwisted's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dektwisted commented|
|-|-|

```
#!html
<pre>
Bug confirmed on NetBSD-current w/ pygtk-2 and gtk-2.4 in addition to Red Hat
Linux 9.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
This is still a problem, however I do not think it is a bug in Twisted.  Gtk
seems not to be releasing the GIL properly.  Adding a print immediately after
the time.sleep() call shows that we are spending about 2 seconds in each sleep,
instead of the request 0.1.

Someone should speak with the pygtk folks about this.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="dektwisted's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dektwisted commented|
|-|-|

```
#!html
<pre>
Jp Calderone [Twisted issue tracker] wrote:

>Jp Calderone &lt;exarkun@...> added the comment:
>
>This is still a problem, however I do not think it is a bug in Twisted.  Gtk
>seems not to be releasing the GIL properly.  Adding a print immediately after
>the time.sleep() call shows that we are spending about 2 seconds in each sleep,
>instead of the request 0.1.
>
>Someone should speak with the pygtk folks about this.
>
>  
>

Thanks for confirming this!  I thought the bug was more or less being 
ignored.  I'm using twisted in a different thread with
a thread safe queue for the time being (this seems to be the 'smart' 
choice even if the gtk reactor is available).

Dave
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Turns out this is a Twisted bug after all.  At some point we stopped
initializing the gtk thread system properly.  Fixed in r13546
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_6"></a>|@markrwilliams commented|
|-|-|

In changeset 2a9091c1fdaceaf3ffbc32edcdb7cb6b953a1bef
```
#!CommitTicketReference repository="" revision="2a9091c1fdaceaf3ffbc32edcdb7cb6b953a1bef"
Merge pull request #552 from scop/8844-scop-http-date-tests

Merge scop/8844-scop-http-date-tests: Add t.w.http.stringToDatetime tests

Author: scop
Reviewer: markrwilliams
Fixes: #8844
```
